pkg/kubelet/pod_workers.go

```diff
@@ -193,42 +206,95 @@ func (p *podWorkers) managePodLoop(podUpdates <-chan UpdatePodOptions) {
-		// the channel is empty, so buffer of size 1 is enough.
-		podUpdates = make(chan UpdatePodOptions, 1)
-		p.podUpdates[uid] = podUpdates
-
-		// Creating a new pod worker either means this is a new pod, or that the
-		// kubelet just restarted. In either case the kubelet is willing to believe
-		// the status of the pod for the first pod worker sync. See corresponding
-		// comment in syncPod.
-		go func() {
-			defer runtime.HandleCrash()
-			p.managePodLoop(podUpdates)
-		}()
-	}
+
 	if !p.isWorking[pod.UID] {
 		p.isWorking[pod.UID] = true
-		podUpdates <- *options
+		p.getWorker(uid) <- *options
+		p.recordDeletingTime(options, now)
+	} else if record, exist := p.recordDeleting[pod.UID]; exist {
+		// If grace period is reduced after stopContainer is requested at least once,
+		// we must reduce the time subtracted from the original request (original grace period - current grace period)
+		// and execute the kill (SIGKILL, not TERM) to the container runtime,
+		// but make sure than no less than 3s have elapsed - so the updated t_kill = t_delete + max(new_grace_period, 3s).
+		// If t_kill > t_now, we should kill the pod immediately.
+		if options.UpdateType == kubetypes.SyncPodKill ||
+			options.UpdateType == kubetypes.SyncPodUpdate {
+			gracePeriod, ok := getGracePeriod(options)
+			if ok && record.GracePeriod > gracePeriod {
+				p.removeWorker(pod.UID)
+				p.isWorking[pod.UID] = true
+				newGracePeriod := gracePeriod
+				if newGracePeriod < shortenedGracePeriodIntervalSeconds {
+					newGracePeriod = shortenedGracePeriodIntervalSeconds
+				}
+				killTime := record.DeleteTime.Add(time.Duration(newGracePeriod) * time.Second)
+				if killTime.Before(now) {
+					killNowOpt := UpdatePodOptions{
+						Pod:            options.Pod,
+						MirrorPod:      options.MirrorPod,
+						UpdateType:     kubetypes.SyncPodKill,
+						OnCompleteFunc: options.OnCompleteFunc,
+						KillPodOptions: options.KillPodOptions,
+					}
+					if killNowOpt.KillPodOptions == nil {
+						killNowOpt.KillPodOptions = &KillPodOptions{}
+					}
+					if killNowOpt.KillPodOptions.PodStatusFunc == nil {
+						killNowOpt.KillPodOptions.PodStatusFunc = func(p *v1.Pod, podStatus *kubecontainer.PodStatus) v1.PodStatus {
+							return options.Pod.Status
+						}
+					}
+					var gracePeriodSecondsOverride int64 = 0
+					killNowOpt.KillPodOptions.PodTerminationGracePeriodSecondsOverride = &gracePeriodSecondsOverride
+
+					p.getWorker(uid) <- killNowOpt
+					p.recordDeletingTime(&killNowOpt, record.DeleteTime)
+				} else {
+					p.getWorker(uid) <- *options
+					p.recordDeletingTime(options, record.DeleteTime)
+				}
+			}
+		}
 	} else {
 		// if a request to kill a pod is pending, we do not let anything overwrite that request.
 		update, found := p.lastUndeliveredWorkUpdate[pod.UID]
 		if !found || update.UpdateType != kubetypes.SyncPodKill {
 			p.lastUndeliveredWorkUpdate[pod.UID] = *options
+			p.recordDeletingTime(options, now)
```

Comment:
SyncPodKill is not the only way that a pod can start terminating - I think we may need to consolidate the SyncPodKill and normal termination paths.